pkg/queue/breaker.go

```diff
@@ -99,60 +137,66 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
-func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
+func (s *Semaphore) Release() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
```

Comment:
should it be `s.reducers = 0`? Since you removed all reducers from capacity, or, rather, `s.capacity--`?

Reply:
We `Release` only a single token, so we need to remove only a single reducer. The number of reducers could be > 1, thus `s.reducers--`.

Agree on the second point - it should be `s.capacity--`.

Reply:
so the final thing should be 

```golang
s.capacity--
s.reducers--
```

?

Reply:
Yep. We get one token back and remove that token from circulation.
We also need a test to make sure this is exactly how it works (we didn't have one before, I guess :pensive: )

Reply:
Updated the code, added the test as well.